pennylane/optimize/adaptive.py

```diff
@@ -0,0 +1,209 @@
+    Largest Gradient: 0.1349349562423238
+
+    Energy: -1.2743971719780331
+    0: ─╭BasisState(M0)─╭G²(0.20)─╭G²(0.19)──────────┤ ╭<𝓗>
+    1: ─├BasisState(M0)─├G²(0.20)─├G²(0.19)─╭G(0.00)─┤ ├<𝓗>
+    2: ─├BasisState(M0)─│─────────├G²(0.19)─│────────┤ ├<𝓗>
+    3: ─├BasisState(M0)─│─────────╰G²(0.19)─╰G(0.00)─┤ ├<𝓗>
+    4: ─├BasisState(M0)─├G²(0.20)────────────────────┤ ├<𝓗>
+    5: ─╰BasisState(M0)─╰G²(0.20)────────────────────┤ ╰<𝓗>
+    Largest Gradient: 0.00040841755397108586
+    """
+
+    def __init__(self, paramopt_steps=10):
+        self.paramopt_steps = paramopt_steps
+
+    @staticmethod
+    def _circuit(params, gates, initial_circuit):
+        """Append parameterized gates to an existing circuit.
+
+        Args:
+            params (array[float]): parameters of the gates to be added
+            gates (list[Operator]): list of the gates to be added
+            initial_circuit (function): user defined circuit that returns an expectation value
+
+        Returns:
+            function: user defined circuit with appended gates
+        """
+        final_circuit = append_gate(params, gates)(initial_circuit)
+
+        return final_circuit()
+
+    def step(self, circuit, operator_pool):
+        r"""Update the circuit with one step of the optimizer.
+
+        Args:
+            circuit (.QNode): user defined circuit returning an expectation value
+            operator_pool (list[Operator]): list of the gates to be used for adaptive optimization
+
+        Returns:
+           .QNode: the optimized circuit
+        """
+        return self.step_and_cost(circuit, operator_pool)[0]
+
+    def step_and_cost(self, circuit, operator_pool, drain_pool=False):
+        r"""Update the circuit with one step of the optimizer and return the corresponding
+        objective function value prior to the step.
+
+        Args:
+            circuit (.QNode): user defined circuit returning an expectation value
+            operator_pool (list[Operator]): list of the gates to be used for adaptive optimization
+            drain_pool (bool): flag to remove selected gates from the operator pool
+
+        Returns:
+            tuple[.QNode, float]: the optimized circuit and the objective function output prior
+            to the step
+        """
+        cost = circuit()
+        device = circuit.device
+
+        if drain_pool:
+            repeated_gates = [
+                gate
+                for gate in operator_pool
+                for operation in circuit.tape.operations
+                if qml.equal(gate, operation, rtol=float("inf"))
+            ]
+            for gate in repeated_gates:
+                operator_pool.remove(gate)
+
+        params = np.array([gate.parameters[0] for gate in operator_pool], requires_grad=True)
+        qnode = qml.QNode(self._circuit, device)
+        grads = qml.grad(qnode)(params, gates=operator_pool, initial_circuit=circuit.func)
+
+        selected_gates = [operator_pool[np.argmax(abs(grads))]]
```

Comment:
Does `argmax` ever return more than one element? Code right now seems to allow for multiple gates being selected, but not sure that ever happens

Reply:
I wrote the code to be flexible. In the original adapt-vqe only one gate is selected at a time. But one can select multiple gates based on a non-zero gradient rule instead of the maximum gradient rule.